### ISSUE TITLE
Replicating performance for ARD on MNV2 using TRADES WRN teacher

### ISSUE THREAD

**Issue by pratik18v** [goldblum/AdversariallyRobustDistillation#2]
Hi,
I am trying to reproduce the performance (as shown in the paper) for ARD on MNV2 using a TRADES WRN teacher. I did the following:

- Train the WRN teacher using the script provided at https://github.com/yaodongyu/TRADES
- Use your code to train the ARD MNV2 model
- For baseline, train MNV2 with the script provided at https://github.com/yaodongyu/TRADES

I am getting the following numbers for a 20-step PGD attack:
**TRADES WRN** 56.43%
**TRADES MNV2** 50.51%
**ARD MNV2 with TRADES WRN teacher** 50.12% 

Do you have any insights on why this might be the case? Or, can you share the scripts you used to generate the baseline numbers?

Thank you

**Comment by goldblum** [comment]
What natural accuracy is your TRADES MNV2 getting?

**Comment by pratik18v** [comment]
It is 77.82%

**Comment by goldblum** [comment]
So you used a different lambda parameter when training TRADES MNV2.  Our ARD model you cited got 82.63% natural accuracy.  The lambda parameter in TRADES trades off natural accuracy for robustness.  If you tune lambda when training TRADES MNV2 to get similar natural accuracy as our ARD mode, you will find that you lose a lot of robust accuracy.

**Comment by pratik18v** [comment]
The model I trained using your code (alpha = 1, temp = 30) yields 78.50% natural accuracy which is comparable to the TRADES MNV2 model (77.82%). However, I noticed that the pre-trained model that you have uploaded yields 82.63% natural accuracy. I trained my model for around 250 epochs. Can you tell me what hyperparameters you used for the uploaded model?

**Comment by goldblum** [comment]
I believe we used the same hyperparameters you're using. I wonder if this is a problem with size averaging or reduction so that the loss is actually weighted slightly differently.  I think PyTorch has changed the defaults for KLDivLoss recently, so it's possible that using a newer version of torch is changing things.

**Comment by pratik18v** [comment]
Thank you for all the help, really appreciate it. Can you please specify the torch version you used, or upload a requirements.txt for the virtualenv you worked in? Thank you!

**Comment by goldblum** [comment]
To be honest, I did this project a long time ago, so I don't remember the exact versions of packages.  I definitely should have made a requirements.txt file.  You can try a few different options of loss averaging.  If loss averaging is the problem here, then you should be able to fix the problem that way.

**Comment by pratik18v** [comment]
Hi,
I tried the using the kl_div loss with all 3 possible aggregation methods (`sum, mean, batchmean`). For `mean` and `batchmean` I am getting similar results (the one i shared previously). For `sum`, the loss explodes so I can't really use it. Looking at the documentation of the previous pytorch versions, I didn't really see any changes in default values of kl_div loss either. Do you have any other suggestions on how to replicate the performance of the pre-trained model that you have uploaded?

**Comment by goldblum** [comment]
I'm not quite sure why you're getting as low natural accuracy as you are.  I have something to try that shouldn't be the issue but is worth a shot:  try doing a 7-step attack during ARD.  I'm pretty sure we did 10-step for our model, but I remember we played around with different numbers of steps.

**Comment by pratik18v** [comment]
I tried that, still the natural accuracy is not crossing 80%. Should I try decreasing alpha?

**Comment by goldblum** [comment]
You can, but we didn't decrease alpha for the experiment you are trying to reproduce.

**Comment by pratik18v** [comment]
Do you have any other suggestions then? Even though I have your pretrained model, I want to be able to similarly train different student teacher pairs for my research.

**Comment by goldblum** [comment]
I don't know why this is happening in your experiments.  I'll try to figure it out, and I'll let you know if I think of something.

**Comment by pratik18v** [comment]
Thank you again for all your help! Also, wanted to ask if you saved the checkpoint which gave the highest natural test accuracy?

**Comment by goldblum** [comment]
No, we did not, but it's possible we saved the model with the highest training accuracy.

**Comment by pratik18v** [comment]
Ok I can try that, save checkpoint based on highest _natural_ train accuracy. I'll let you know if that helps.

**Comment by pratik18v** [comment]
I re-ran the training scripts for both ARD and TRADES (beta = 3), this time around saving model corresponding to highest natural _test_ accuracy. The following are the results for a 20-step PGD attack:

**TRADES MNV2**
No Attack: 81.68%
PGD: 47.40%

**ARD MNV2**
No Attack: 80.29%
PGD: 45.42%

I had the following question regarding the first potential issue you raised reading loss aggregating methods. Since we're using alpha = 1, that implies we're only minimizing one loss term during training. It seems like if there aren't multiple loss terms involved (of varying scales), loss aggregation method should not matter (as long as loss magnitude is not very small resulting in vanishing gradients). Do you remember what the order of magnitude of the KL loss term (without multiplying with alpha and temp^2) was while training? I am observing small magnitudes of the order approx 10^-3 in the beginning, and approx 10^-4 towards the end.

**Comment by goldblum** [comment]
There are multiple loss terms in the sense of data in a batch.  Choice of aggregation would scale the loss and effectively change the learning rate and thus potentially effect performance.  I do not remember loss magnitute.  BTW, in the test you just ran, how many epochs did it take to get to those results?

**Comment by pratik18v** [comment]
Wouldn't an aggregation method that averages the loss per batch take care of varying loss magnitudes between batches (if that's what you meant)?
For the TRADES model the last epoch saved was at 104 (ran for 125). For the ARD model, it was epoch 185 (ran for 250).

**Comment by goldblum** [comment]
That's not what I mean.  I'm saying if you average over the batch or divide by the number of network outputs or don't average at all, you will get different results.

**Comment by pratik18v** [comment]
So far based on the tests I've ran, averaging over the batch or dividing by the number of network outputs is giving approximately similar results. And using summing just causes the loss to explode. I didn't change anything else in the code other than the aggregation method for the kldiv loss.

**Comment by pratik18v** [comment]
Hi, I think I have figured out the issue. I was earlier training the models using a smaller batch size (64) as I was working with a smaller GPU. I re ran the training using the default batch size of 128 and following are the results:

**ARD MNV2**
| | My Model | Author Model |
| -- | -- | -- |
| **No Attack** | 82.48% | 82.63% |
| **PGD** | 49.64% | 50.98% |

The issue with the natural accuracy being low seems to be resolved, as my model and the model you uploaded have approximately equal natural accuracies. Even the adversarial accuracies for our models seem to be close enough.

**TRADES MNV2**
|  | beta = 6 | beta = 3 | beta = 2 |
| -- | -- | -- | -- |
| **No Attack** | 81.29% | 83.81% | 85.54% |
| **PGD** | 51.44% | 49.37% | 46.93% |

The issue that remains is with the adversarial accuracy for the TRADES MNV2 model. As you can see from the table above, the most comparable model (in terms of equivalent natural accuracies) is the one corresponding to beta = 3. However, its adversarial accuracy is considerably higher that the TRADES MNV2 model reported in your paper (49.37% vs. 44.79%). As you can also see from the table above, if I reduce the beta any further I am getting much higher natural accuracy so it seems like beta = 3 will remain the most comparable model corresponding to alpha = 1 (and temp = 30) model for ARD.
Did you also use the [TRADES code](https://github.com/yaodongyu/TRADES) provided by the authors?
Please let me know if you have any clues regarding this particular discrepancy.
Thank you for your time and all the help so far.
Regards

**Comment by zjysteven** [comment]
@pratik18v I'm observing the same problem here, namely the ARD MNv2 is actually less robust than TRADES MNv2 when using a WRN34-10 teacher model. I was suspecting this might be due to the huge mismatch between the capacity of the student and the teacher (which has been shown can make distillation ineffective), so I also trained an ARD MNv2 using ResNet18 as the teacher. However, I still observe lower robustness than the TRADES counterpart.
